src/modules/screen_plugins/wl.c

```diff
@@ -1,200 +1,246 @@
+                          uint32_t name, const char *interface,
+                          uint32_t version) {
+    struct cl_display *display = (struct cl_display *)data;
+    struct cl_output *output = calloc(1, sizeof(struct cl_output));
+    output->display = display;
+    if (strcmp(interface, wl_output_interface.name) == 0) {
+        output->wl_output =
+            wl_registry_bind(registry, name, &wl_output_interface, 4);
+        wl_output_add_listener(output->wl_output, &wl_output_listener, output);
+        output->name = NULL;
+        wl_list_insert(&display->outputs, &output->link);
+    } else if (strcmp(interface, wl_shm_interface.name) == 0) {
+        display->shm = wl_registry_bind(registry, name, &wl_shm_interface, 1);
+    } else if (strcmp(interface, zwlr_screencopy_manager_v1_interface.name) ==
+               0) {
+        display->screencopy_manager = wl_registry_bind(
+            registry, name, &zwlr_screencopy_manager_v1_interface, 3);
     }
 }
-
-static void handle_global_remove(void *data, struct wl_registry *registry, uint32_t name) {
-    
+static const struct wl_registry_listener registry_listener = {
+    .global = handle_global,
+    .global_remove = noop,
+};
+static void wl_cleanup(struct cl_display *display) {
+    struct cl_output *output;
+    struct cl_output *tmp_output;
+    wl_list_for_each_safe(output, tmp_output, &display->outputs, link) {
+        output->frame->copy_done = false;
+        output->frame->copy_err = false;
+        if (output->screencopy_frame != NULL) {
+            zwlr_screencopy_frame_v1_destroy(output->screencopy_frame);
+        }
+        if (output->buffer->shm_data) {
+            munmap(output->buffer->shm_data, output->frame->size);
+        }
+        if (output->buffer->wl_buffer) {
+            wl_buffer_destroy(output->buffer->wl_buffer);
+        }
+        free(output->buffer);
+        wl_output_destroy(output->wl_output);
+        wl_list_remove(&output->link);
+        free(output->name);
+        free(output);
+    }
+    if (display->wl_registry) {
+        wl_registry_destroy(display->wl_registry);
+    }
+    if (display->screencopy_manager) {
+        zwlr_screencopy_manager_v1_destroy(display->screencopy_manager);
+    }
+    wl_display_flush(display->wl_display);
+    wl_display_disconnect(display->wl_display);
 }
-
-static void dtor(void) {
-    buffer_copy_done = false;
-    buffer_copy_err = false;
-    
-    /* Free everything */
-    if (buffer.wl_buffer) {
-        wl_buffer_destroy(buffer.wl_buffer);
-        buffer.wl_buffer = NULL;
-    }
-    if (buffer.data) {
-        munmap(buffer.data, (size_t)buffer.height * buffer.stride);
-        buffer.data = NULL;
+static int get_frame_brightness(const char *id, const char *env) {
+    struct cl_display display = {};
+    struct cl_output *output;
+    int ret = 0;
+    char *socket_file;
+    int len = snprintf(NULL, 0, "%s/%s", env, id);
```

Comment:
You should use `fetch_wl_display` utility: if the display is already opened by some other module (eg: gamma), you will reuse the same wl_display.